journal.py

```diff
@@ -91,43 +93,49 @@ def __repr__(self):
 # all the views
 
 
-@view_config(route_name='add', request_method='POST')
-def add_entry(request):
-    title = request.params.get('title')
-    text = request.params.get('text')
-    Entry.write(title=title, text=text)
-    return HTTPFound(request.route_url('home'))
-
-
-@view_config(route_name='add', renderer="templates/add.jinja2")
+@view_config(route_name='add', xhr=True, renderer='json')
+@view_config(route_name='add', xhr=False, renderer="templates/add.jinja2")
 def add_entry_view(request):
-    username = request.params.get('username', '')
-    return {'username': username}
+    if not request.authenticated_userid:
+        return HTTPFound(request.route_url('login'))
 
+    if request.method == 'POST':
+        title = request.params.get('title')
+        text = request.params.get('text')
+        Entry.write(title=title, text=text)
+        return HTTPFound(request.route_url('home'))
+    elif request.method == 'GET':
+        return {}
 
-@view_config(route_name='edit', request_method='POST')
-def edit_entry(request):
-    entry = Entry.id_lookup(request.matchdict['id'])
-    title = request.params.get('title')
-    text = request.params.get('text')
-    id = entry.id
-    Entry.edit(title=title, text=text, id=id)
-    return HTTPFound(request.route_url('home'))
 
+@view_config(route_name='edit', xhr=True, renderer='json')
+@view_config(route_name='edit', xhr=False, renderer='templates/edit.jinja2')
+def edit_entry(request):
+    if not request.authenticated_userid:
+        return HTTPFound(request.route_url('login'))
 
-@view_config(route_name='edit', renderer="templates/edit.jinja2")
-def edit_view(request):
-    entry = Entry.id_lookup(request.matchdict['id'])
-    return {'entry': {
-            'id': entry.id,
-            'title': entry.title,
-            'text': entry.text,
-            'created': entry.created}}
+    id = request.matchdict['id']
+    if request.method == 'POST':
+        entry = Entry.id_lookup(request.matchdict['id'])
+        title = request.params.get('title')
+        text = request.params.get('text')
+        id = entry.id
+        Entry.edit(title=title, text=text, id=id)
+        entry = Entry.id_lookup(id)
```

Comment:
afaict, line 124 here is an exact duplicate of line 119.  Neither of those is necessary if your `edit` method returns the entry it just edited.  Does it?